include/dxc/Support/WinAdapter.h

```diff
@@ -1039,8 +1040,60 @@ class CHandle {
   HANDLE m_h;
 };
 
+
+/////////////////////////////////////////////////////////////////////////////
+// CComBSTR
+
+class CComBSTR
+{
+public:
+    BSTR m_str;
+    CComBSTR() : m_str(nullptr) {};
+    CComBSTR(_In_ int nSize, LPCWSTR sz);
+    ~CComBSTR() throw() {
+      SysFreeString(m_str);
+    }
+
+    operator BSTR() const throw()
+    {
+        return m_str;
+    }
+
+    bool operator==(_In_ const CComBSTR& bstrSrc) const throw();
+
+    BSTR* operator&() throw()
+    {
+        return &m_str;
+    }
+
+    BSTR Detach() throw()
+    {
+        BSTR s = m_str;
+        m_str = NULL;
+        return s;
+    }
+
+};
+
+
 #endif // __cplusplus
 
 #endif // _WIN32
 
+#ifdef __cplusplus
+
+#include <string>
+#include <vector>
+//===--------- Convert argv to wchar ----------------===//
+class WArgV {
+  std::vector<std::wstring> WStringVector;
+  std::vector<const wchar_t *> WCharPtrVector;
+
+public:
+  WArgV(int argc, const char **argv);
+  WArgV(int argc, const wchar_t **argv);
+  const wchar_t **argv() { return WCharPtrVector.data();}
+};
+#endif
+
```

Comment:
I have looked at the places that `WArgV` is used. They convert args to wchar that are then passed either to the `MainArgs` constructor or to `RewriteWithOptions`. The `MainArgs` constructor can already handle either wchar or char arguemtns and all that `RewriteWithOptions` does is pass it into the `MainArgs` constructor.

I think this is redundant. I think we should change `RewriteWithOptions` to take a `MainArgs` object and make use of the existing code to perform this conversion of arguments. If that code can be improved, I think it should be improved there. https://github.com/microsoft/DirectXShaderCompiler/blob/60530e57ebdcc5833fce48313286f90803b05bb5/lib/DxcSupport/HLSLOptions.cpp#L168